File: android/app/src/main/java/ca/popuphub/app/WidgetBridgePlugin.java
package ca.popuphub.app;

import android.content.Context;
import android.content.Intent;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import ca.popuphub.app.widget.PopupHubWidgetProvider;

@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {

    public static final String PREFS_NAME = "popup_hub_widget";
    public static final String TOKEN_KEY = "widget_auth_token";
    public static final String SNAPSHOT_KEY = "widget_snapshot_json";

    @PluginMethod
    public void save(PluginCall call) {
        String token = call.getString("token");
        String snapshotJson = call.getString("snapshotJson");
        if (token == null || snapshotJson == null) {
            call.reject("token and snapshotJson required");
            return;
        }

        Context context = getContext();
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(TOKEN_KEY, token)
            .putString(SNAPSHOT_KEY, snapshotJson)
            .apply();

        Intent updateIntent = new Intent(context, PopupHubWidgetProvider.class);
        updateIntent.setAction(PopupHubWidgetProvider.ACTION_REFRESH);
        context.sendBroadcast(updateIntent);

        call.resolve(new JSObject());
    }

    @PluginMethod
    public void clear(PluginCall call) {
        Context context = getContext();
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(TOKEN_KEY)
            .remove(SNAPSHOT_KEY)
            .apply();

        Intent updateIntent = new Intent(context, PopupHubWidgetProvider.class);
        updateIntent.setAction(PopupHubWidgetProvider.ACTION_REFRESH);
        context.sendBroadcast(updateIntent);

        call.resolve(new JSObject());
    }
}

File: ios/App/App/WidgetBridge/WidgetBridgePlugin.swift
import Foundation
import Capacitor
import WidgetKit

enum WidgetSharedKeys {
    static let appGroup = "group.ca.popuphub.app"
    static let tokenKey = "widget_auth_token"
    static let snapshotKey = "widget_snapshot_json"
}

@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetBridgePlugin"
    public let jsName = "WidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "save", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
    ]

    @objc func save(_ call: CAPPluginCall) {
        guard let token = call.getString("token"),
              let snapshotJson = call.getString("snapshotJson") else {
            call.reject("token and snapshotJson required")
            return
        }

        let defaults = UserDefaults(suiteName: WidgetSharedKeys.appGroup)
        defaults?.set(token, forKey: WidgetSharedKeys.tokenKey)
        defaults?.set(snapshotJson, forKey: WidgetSharedKeys.snapshotKey)

        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve()
    }

    @objc func clear(_ call: CAPPluginCall) {
        let defaults = UserDefaults(suiteName: WidgetSharedKeys.appGroup)
        defaults?.removeObject(forKey: WidgetSharedKeys.tokenKey)
        defaults?.removeObject(forKey: WidgetSharedKeys.snapshotKey)

        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve()
    }
}
